enhancements/sha-bundle-digest.md

```diff
@@ -0,0 +1,80 @@
+
+This set of changes will allow SDK users in air-gapped (disconnected) environments to generate bundles with the image digest. This change is necessary due to non-deterministic behavior with images identified via tags, this issue specifically occurs in the OCP registry due to how the digest is generated with tag name as inputs.
+
+### Goals
+
+-   Give users the ability to the enable digest-based image resolution via a CLI flag when generating bundles.
+-   Allow users to specify additional images to resolve digests via an environment variable.
+-   Add the ability to resolve image digests from image tag dynamically using local docker registry.
+-   OPEN QUESTION: Store digest and tag mappings in file so that user-written code can resolve tags to digest.
+
+### Non-Goals
+
+-   Migrate existing `operator-sdk` based operators from image tags to image digests.
+-   Identify locations in user code that reference images by tag.
+
+## Proposal
+
+-   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. 
+
+-   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 
+
+
+### Risks and Mitigations
+
+Since the feature is opt-in, it carries the risk of the end-user enabling it without a correct understanding of the feature and its implications. This can be mitigated with clear documentation both on the website and in the CLI help section.
+
+## Design Details
+
+ -   Add a `--image-digests` flag to the `operator-sdk generate bundle` command that will enable this functionality.
+ -   Add a variable to the generated `Makefile` called `USE_IMAGE_DIGESTS` that defaults to `false` and when enabled, it passes the `--use-image-digests` flag to `operator-sdk generate bundle`.
+ -   When `operator-sdk generate-bundle` is called with the `--use-image-digests` flag enabled, it will execute the following logic.
+ 	1. Search the environment that is defined in the `manager.yaml` file for all variables that begin with `RELATED_IMAGE_` to build a list of images that need to have tags resolved to digests along with the Operator’s main image and rbac-proxy image.
+ 	 1.  For each image in the list, it will use `operator-manifest-tools` [https://github.com/operator-framework/operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools) to resolve the image tag to a digest
+ 	 1.  When building the cluster CSV, all images that have been identified as needing a digest instead of a tag should have that  digest as an annotation in the CSV.
```

Comment:
This section should focus on the implementation details that the engineer will have to take into account when building out a feature that does what is defined in the proposal. For example:

```
- As of this EP, the new `--use-image-digests` is the only flag to pass to the `operator-sdk generate bundle` command. In order to allow users full control over the command's flags, the `Makefile` should use an intermediary variable to hold the flags that a user can overwrite entirely. For example:
\```
GENERATE_BUNDLE_FLAGS ?= ""
USE_IMAGE_DIGESTS ?= false
ifeq ($(GENERATE_BUNDLE_FLAGS),"")
    ifeq($(USE_IMAGE_DIGESTS),true)
        GENERATE_BUNDLE_FLAGS += "--use-image-digests"
    endif
endif
bundle:
    operator-sdk generate bundle $(GENERATE_BUNDLE_FLAGS)
\```
- The `RELATED_IMAGE_*` variables should be defined in the `manager.yaml` file as outlined [here](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions).
- In order to resolve digests from tags, the CLI will use the `imageresolver` package in [operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools).
- The output cluster CSV will have the `relatedImage` annotation that is explained in the [OpenShift Documentation](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs).
```

Now the details directly tie into specific proposal items, which themselves tie into specific goals.
Also the Make code is just written by hand now, verify it if you opt to add it into this proposal.